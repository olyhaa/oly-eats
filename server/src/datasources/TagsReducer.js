export const tagReducer = (tag) => {
  if (!tag) {
    return null;
  }
  return {
    id: tag.id,
    typeid: tag.typeid,
    label: tag.label,
  };
};

export const tagMutationReducer = ({
  success = false,
  message = undefined,
  tag = null,
}) => {
  return {
    success,
    message,
    tag: tagReducer(tag),
  };
};
