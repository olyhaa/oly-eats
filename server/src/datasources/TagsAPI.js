import { DataSource } from 'apollo-datasource';
import { tagReducer, tagMutationReducer } from './TagsReducer';

class TagsAPI extends DataSource {
  constructor({ store }) {
    super();
    this.store = store;
  }

  async getAllTags({ typeid }) {
    const tagType = await this.store.TagType.findByPk(typeid);
    if (!tagType) {
      return [];
    }
    const tagArray = await this.store.Tag.findAll({
      where: { tagTypeId: tagType.id },
    });
    return Array.isArray(tagArray)
      ? tagArray.map((tag) => tagReducer(tag))
      : [];
  }

  async addTag({ typeid, label }) {
    const tagType = await this.store.TagType.findByPk(typeid);
    if (!tagType) {
      return tagMutationReducer({
        success: false,
        message: 'Type ID does not exist',
      });
    }
    const tag = await this.store.Tag.create({
      tagTypeId: tagType.id,
      label,
    });

    return tagMutationReducer({
      success: true,
      tag,
    });
  }

  async deleteTag({ id }) {
    const tag = await this.store.Tag.findByPk(id);
    if (!tag) {
      return tagMutationReducer({
        success: false,
        message: 'ID not found',
      });
    }
    await tag.destroy();
    return tagMutationReducer({
      success: true,
    });
  }

  async updateTag({ id, label }) {
    let tag = await this.store.Tag.findByPk(id);
    if (!tag) {
      return tagMutationReducer({
        success: false,
        message: 'ID does not exist',
      });
    }
    await tag.update({ label });
    tag = await this.store.Tag.findByPk(id);
    return tagMutationReducer({ tag, success: true });
  }
}

export default TagsAPI;
