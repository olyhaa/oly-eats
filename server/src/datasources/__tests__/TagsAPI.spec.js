import TagsAPI from '../TagsAPI';
import { tagReducer } from '../TagsReducer';

const mockTagTypes = [
  {
    id: '1',
    label: 'category',
  },
  {
    id: '2',
    label: 'cuisine',
  },
];

const mockTags = [
  {
    id: '1',
    typeid: '1',
    label: 'bakery',
  },
  {
    id: '2',
    typeid: '1',
    label: 'casserole',
  },
  {
    id: '3',
    typeid: '1',
    label: 'salad',
  },
  {
    id: '7',
    typeid: '2',
    label: 'american',
  },
  {
    id: '8',
    typeid: '2',
    label: 'asian',
  },
];

const mockStore = {
  TagType: {
    findByPk: jest.fn(),
  },
  Tag: {
    create: jest.fn(),
    findAll: jest.fn(),
    findByPk: jest.fn(),
  },
};
const tagsDatasource = new TagsAPI({ store: mockStore });

describe('getAllTags', () => {
  it('returns empty array for empty db for valid typeid', async () => {
    mockStore.TagType.findByPk.mockReturnValueOnce(mockTagTypes[0]);
    mockStore.Tag.findAll.mockReturnValueOnce([]);

    const response = await tagsDatasource.getAllTags({
      typeid: mockTagTypes[0].id,
    });
    expect(mockStore.TagType.findByPk).toBeCalledWith(mockTagTypes[0].id);
    expect(mockStore.Tag.findAll).toBeCalledWith({
      where: { typeid: mockTagTypes[0].id },
    });
    expect(response).toEqual([]);
  });

  it('returns empty array for empty db for bad typeid', async () => {
    mockStore.TagType.findByPk.mockReturnValueOnce(null);
    const bad_id = 'bad id';

    const response = await tagsDatasource.getAllTags({
      typeid: bad_id,
    });

    expect(mockStore.TagType.findByPk).toBeCalledWith(bad_id);
    expect(response).toEqual([]);
  });

  it('returns array of tags for valid typeid', async () => {
    mockStore.Tag.findAll.mockReturnValueOnce(mockTags.slice(0, 3));
    mockStore.TagType.findByPk.mockReturnValueOnce(mockTagTypes[0]);

    const response = await tagsDatasource.getAllTags({
      typeid: mockTagTypes[0].id,
    });
    expect(mockStore.Tag.findAll).toBeCalledWith({
      where: { typeid: mockTagTypes[0].id },
    });
    expect(response).toEqual(
      mockTags.slice(0, 3).map((tag) => tagReducer(tag))
    );
  });
});

describe('addTag', () => {
  it('calls store creator and returns result - valid typeid', async () => {
    const test_label = 'test label';
    mockStore.TagType.findByPk.mockReturnValueOnce(mockTagTypes[0]);
    mockStore.Tag.create.mockReturnValueOnce(mockTags[0]);

    // check the result of the fn
    const res = await tagsDatasource.addTag({
      typeid: mockTagTypes[0].id,
      label: test_label,
    });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.TagType.findByPk).toBeCalledWith(mockTagTypes[0].id);
    expect(mockStore.Tag.create).toBeCalledWith({
      typeid: mockTagTypes[0].id,
      label: test_label,
    });
  });

  it('does not call store creator and returns result - invalid typeid', async () => {
    const test_label = 'test label';
    const bad_type_id = 'bad id';
    mockStore.TagType.findByPk.mockReturnValueOnce(null);

    // check the result of the fn
    const res = await tagsDatasource.addTag({
      typeid: bad_type_id,
      label: test_label,
    });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.TagType.findByPk).toBeCalledWith(bad_type_id);
  });
});

describe('deleteTag', () => {
  it('calls store destroy and returns result - id exists', async () => {
    const findPkVal = Object.assign({}, mockTags[2]);
    findPkVal.destroy = jest.fn();

    mockStore.Tag.findByPk.mockReturnValueOnce(findPkVal);
    findPkVal.destroy.mockReturnValueOnce('done');

    // check the result of the fn
    const res = await tagsDatasource.deleteTag({ id: findPkVal.id });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.Tag.findByPk).toBeCalledWith(findPkVal.id);
    expect(findPkVal.destroy).toBeCalled();
  });

  it('calls store destroy and returns result - id not found', async () => {
    mockStore.Tag.findByPk.mockReturnValueOnce(null);
    const bad_id = 'bad id';

    // check the result of the fn
    const res = await tagsDatasource.deleteTag({ id: bad_id });
    expect(res).toMatchSnapshot();

    expect(mockStore.Tag.findByPk).toBeCalledWith(bad_id);
  });
});

describe('updateTag', () => {
  it('calls store update with values and returns result - id exists', async () => {
    const findPkVal = Object.assign({}, mockTags[1]);
    const updatedPkVal = Object.assign({}, findPkVal);
    updatedPkVal.label = 'new label';

    findPkVal.update = jest.fn();

    mockStore.Tag.findByPk
      .mockReturnValueOnce(findPkVal)
      .mockReturnValueOnce(updatedPkVal);
    findPkVal.update.mockReturnValueOnce(updatedPkVal);
    const new_label = 'new label';

    // check the result of the fn
    const res = await tagsDatasource.updateTag({
      id: findPkVal.id,
      label: new_label,
    });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.Tag.findByPk).toBeCalledWith(findPkVal.id);
    expect(findPkVal.update).toBeCalledWith({ label: new_label });
  });

  it('calls store update and returns result - id not found', async () => {
    mockStore.Tag.findByPk.mockReturnValueOnce(null);
    const bad_id = 'bad_id';

    // check the result of the fn
    const res = await tagsDatasource.updateTag({
      id: bad_id,
      label: 'new label',
    });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.Tag.findByPk).toBeCalledWith(bad_id);
  });
});
