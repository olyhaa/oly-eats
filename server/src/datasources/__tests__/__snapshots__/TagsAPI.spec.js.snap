// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTag calls store creator and returns result - valid typeid 1`] = `
Object {
  "message": undefined,
  "success": true,
  "tag": Object {
    "id": "1",
    "label": "bakery",
    "typeid": "1",
  },
}
`;

exports[`addTag does not call store creator and returns result - invalid typeid 1`] = `
Object {
  "message": "Type ID does not exist",
  "success": false,
  "tag": null,
}
`;

exports[`deleteTag calls store destroy and returns result - id exists 1`] = `
Object {
  "message": undefined,
  "success": true,
  "tag": null,
}
`;

exports[`deleteTag calls store destroy and returns result - id not found 1`] = `
Object {
  "message": "ID not found",
  "success": false,
  "tag": null,
}
`;

exports[`updateTag calls store update and returns result - id not found 1`] = `
Object {
  "message": "ID does not exist",
  "success": false,
  "tag": null,
}
`;

exports[`updateTag calls store update with values and returns result - id exists 1`] = `
Object {
  "message": undefined,
  "success": true,
  "tag": Object {
    "id": "2",
    "label": "new label",
    "typeid": "1",
  },
}
`;
